File: wrappers/dotnet/indy-sdk-dotnet-test/AnonCredsTests/ProverCreateProofTest.cs
using Hyperledger.Indy.AnonCredsApi;
using Hyperledger.Indy.WalletApi;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Threading.Tasks;

namespace Hyperledger.Indy.Test.AnonCredsTests
{
    [TestClass]
    public class ProverCreateProofTest : AnonCredsIntegrationTestBase
    {
        private readonly string requestedCredentialsJson = string.Format("{{" +
            "\"self_attested_attributes\":{{}}," +
            "\"requested_attributes\":{{\"attr1_referent\":{{\"cred_id\":\"{0}\", \"revealed\":true}}}}," +
            "\"requested_predicates\":{{\"predicate1_referent\":{{\"cred_id\":\"{1}\"}}}}" +
            "}}", credentialId1, credentialId1);

        [TestMethod]
        public async Task TestProverCreateProofWorks()
        {
            var schemasJson = new JObject(new JProperty(gvtSchemaId, JObject.Parse(gvtSchema))).ToString();
            var credentialDefsJson = new JObject(new JProperty(issuer1gvtCredDefId, JObject.Parse(issuer1gvtCredDef))).ToString();
            var revocStatesJson = "{}";

            var proofJson = await AnonCreds.ProverCreateProofAsync(
                wallet, 
                proofRequest, 
                requestedCredentialsJson,
                masterSecretId, 
                schemasJson, 
                credentialDefsJson, 
                revocStatesJson);

            Assert.IsNotNull(proofJson);
        }
        
        [TestMethod]
        public async Task TestProverCreateProofWorksForUsingNotSatisfyClaim()
        {
            var requestedCredentialsJson = string.Format("{{\"self_attested_attributes\":{{}},\n" +
                 "                                    \"requested_attributes\":{{\"attr1_referent\":{{\"cred_id\":\"{0}\", \"revealed\":true}}}},\n" +
                 "                                    \"requested_predicates\":{{}}\n" +
                 "                                   }}", credentialId2);

            var schemasJson = new JObject(new JProperty(xyzSchemaId, xyzSchema)).ToString();
            var credentialDefsJson = new JObject(new JProperty(issuer1xyzCredDef, issuer1xyzCredDef)).ToString();
            var revocStatesJson = "{}";

            await Assert.ThrowsExceptionAsync<InvalidStructureException>(() =>
                AnonCreds.ProverCreateProofAsync(wallet, proofRequest, requestedCredentialsJson, 
                    masterSecretId, schemasJson, credentialDefsJson, revocStatesJson)
            );
        }

        [TestMethod]
        public async Task TestProverCreateProofWorksForInvalidMasterSecret()
        {
            var schemasJson = new JObject(new JProperty(gvtSchemaId, JObject.Parse(gvtSchema))).ToString();
            var credentialDefsJson = new JObject(new JProperty(issuer1gvtCredDefId, JObject.Parse(issuer1gvtCredDef))).ToString();
            var revocStatesJson = "{}";
            
            var ex = await Assert.ThrowsExceptionAsync<WalletItemNotFoundException>(() =>
                AnonCreds.ProverCreateProofAsync(wallet, proofRequest, requestedCredentialsJson, "wrong_master_secret", schemasJson, credentialDefsJson, revocStatesJson)
            );
        }

        [TestMethod]
        public async Task TestProverCreateProofWorksForInvalidSchemas()
        {
            var schemasJson = "{}";
            var credentialDefsJson = new JObject(new JProperty(issuer1gvtCredDefId, issuer1gvtCredDef)).ToString();
            var revocStatesJson = "{}";

            var ex = await Assert.ThrowsExceptionAsync<InvalidStructureException>(() =>
                AnonCreds.ProverCreateProofAsync(wallet, proofRequest, requestedCredentialsJson, masterSecretId, schemasJson, credentialDefsJson, revocStatesJson)

            );
        }

        [TestMethod]
        public async Task TestProverCreateProofWorksForInvalidRequestedClaimsJson()
        {
            var schemasJson = new JObject(new JProperty(gvtSchemaId, gvtSchema)).ToString();
            var credentialDefsJson = new JObject(new JProperty(issuer1gvtCredDefId, issuer1gvtCredDef)).ToString();
            var revocStatesJson = "{}";
            var requestedCredentialsJson = JsonConvert.SerializeObject(
                new { self_attested_attributes = new JObject(), requested_predicates = new JObject() }
            );


            var ex = await Assert.ThrowsExceptionAsync<InvalidStructureException>(() =>
                AnonCreds.ProverCreateProofAsync(wallet, proofRequest, requestedCredentialsJson, masterSecretId, schemasJson, credentialDefsJson, revocStatesJson)

            );
        }

    }
}

File: wrappers/java/src/test/java/org/hyperledger/indy/sdk/anoncreds/ProverCreateProofTest.java
package org.hyperledger.indy.sdk.anoncreds;

import org.hyperledger.indy.sdk.InvalidStructureException;
import org.hyperledger.indy.sdk.wallet.WalletItemNotFoundException;
import org.junit.Test;
import org.json.JSONObject;
import java.util.concurrent.ExecutionException;

import static org.hamcrest.CoreMatchers.isA;
import static org.junit.Assert.assertNotNull;

public class ProverCreateProofTest extends AnoncredsIntegrationTest {

	private String requestedCredentialsJson = String.format("{" +
			"\"self_attested_attributes\":{}," +
			"\"requested_attributes\":{\"attr1_referent\":{\"cred_id\":\"%s\", \"revealed\":true}}," +
			"\"requested_predicates\":{\"predicate1_referent\":{\"cred_id\":\"%s\"}}" +
			"}", credentialId1, credentialId1);

	@Test
	public void testProverCreateProofWorks() throws Exception {

		String schemasJson = new JSONObject().put(gvtSchemaId, new JSONObject(gvtSchema)).toString();
		String credentialDefsJson = new JSONObject().put(issuer1gvtCredDefId, new JSONObject(issuer1gvtCredDef)).toString();
		String revocStatesJson = new JSONObject().toString();

		String proofJson = Anoncreds.proverCreateProof(wallet, proofRequest, new JSONObject(requestedCredentialsJson).toString(),
				masterSecretId, schemasJson, credentialDefsJson, revocStatesJson).get();
		assertNotNull(proofJson);
	}

	@Test
	public void testProverCreateProofWorksForUsingNotSatisfyCredential() throws Exception {

		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(InvalidStructureException.class));

		String requestedCredentialsJson = String.format("{\"self_attested_attributes\":{},\n" +
				"                                    \"requested_attributes\":{\"attr1_referent\":{\"cred_id\":\"%s\", \"revealed\":true}},\n" +
				"                                    \"requested_predicates\":{}\n" +
				"                                   }", credentialId2);

		String schemasJson = new JSONObject().put(xyzSchemaId, xyzSchema).toString();
		String credentialDefsJson = new JSONObject().put(issuer1xyzCredDef, issuer1xyzCredDef).toString();
		String revocStatesJson = new JSONObject().toString();

		Anoncreds.proverCreateProof(wallet, proofRequest, new JSONObject(requestedCredentialsJson).toString(),
				masterSecretId, schemasJson, credentialDefsJson, revocStatesJson).get();
	}

	@Test
	public void testProverCreateProofWorksForInvalidMasterSecret() throws Exception {

		String schemasJson = new JSONObject().put(gvtSchemaId, new JSONObject(gvtSchema)).toString();
		String credentialDefsJson = new JSONObject().put(issuer1gvtCredDefId, new JSONObject(issuer1gvtCredDef)).toString();
		String revocStatesJson = new JSONObject().toString();

		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(WalletItemNotFoundException.class));

		Anoncreds.proverCreateProof(wallet, proofRequest, new JSONObject(requestedCredentialsJson).toString(),
				"wrong_master_secret", schemasJson, credentialDefsJson, revocStatesJson).get();
	}

	@Test
	public void testProverCreateProofWorksForInvalidSchemas() throws Exception {

		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(InvalidStructureException.class));

		String schemasJson = new JSONObject().toString();
		String credentialDefsJson = new JSONObject().put(issuer1gvtCredDefId, issuer1gvtCredDef).toString();
		String revocStatesJson = new JSONObject().toString();

		Anoncreds.proverCreateProof(wallet, proofRequest, new JSONObject(requestedCredentialsJson).toString(),
				masterSecretId, schemasJson, credentialDefsJson, revocStatesJson).get();
	}

	@Test
	public void testProverCreateProofWorksForInvalidRequestedCredentialsJson() throws Exception {

		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(InvalidStructureException.class));

		String schemasJson = new JSONObject().put(gvtSchemaId, gvtSchema).toString();
		String credentialDefsJson = new JSONObject().put(issuer1gvtCredDefId, issuer1gvtCredDef).toString();
		String revocStatesJson = new JSONObject().toString();
		String requestedCredentialsJson = new JSONObject().put("self_attested_attributes", new JSONObject()).
														   put("requested_predicates", new JSONObject()).toString();


		/*String requestedCredentialsJson = "{\"self_attested_attributes\":{},\n" +
				"                      \"requested_predicates\":{}\n" +
				"                    }";*/

		Anoncreds.proverCreateProof(wallet, proofRequest, requestedCredentialsJson, masterSecretId, schemasJson, credentialDefsJson, revocStatesJson).get();
	}
}
